src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
```

Comment:
This would make the field nullable which is a good movement.

Reply:
I don't think that this would be a good movement.
It would unnecessarily add more logic to the wrapper.
We'd need a null check on every access to the wrapped `CancellationTokenSource`, even in cases where we know for sure that it can't be null.
We can always make the whole wrapper nullable.